# Project 1: Traveling Salesperson Problem (TSP)<br><small>Joshua Nasiatka (2016.11.13)

## Problem 1: 
This problem solves a real-world Traveling Salesperson Problem (TSP) using distance data. We will find the shortest total path between the Massachusetts cities/towns: Boston, Fall River, Fitchburg, Lowell, North Adams, Springfield and Worcester. Follow these steps to do this:  

**( a ) Use the brute force algorithm that goes through each path to determine the shortest total distance. Find the distance and write down the path as it goes through each town. Start and end with Fitchburg and write down the town names of the shortest path.**

In [1]:
push!(LOAD_PATH,".");

In [2]:
using TSP

I cannot fathom how to go about the round-trip Fitchburg => (cities) => Fitchburg, I'm probably missing something simple, I just can't figure out what it is. This goes for most of the questions; I'm not sure how to explicitly check for Fitchburg as Starting position/location.

In [3]:
min_dist, tpath, dist_map = run_TSP("map_q1.json");

The Shortest Distance and Position is: (596457.0,2930)


The path of which is: 
String["North Adams, MA 01247, USA","Fitchburg, MA 01420, USA","Lowell, MA, USA","Boston, MA, USA","Fall River, MA, USA","Worcester, MA, USA","Springfield, MA, USA","North Adams, MA 01247, USA"]

In [4]:
0.000621371*dist_map

7×7 Array{Float64,2}:
  0.0      47.0539   83.0232   31.4153   87.2026   78.7066   27.2366
 52.9017    0.0      51.8894   30.3484  136.016    91.1992   47.1925
 84.1908   52.3747    0.0      77.0351  169.439   102.106    56.9598
 33.1967   30.121    77.0469    0.0     116.311    93.7052   42.2358
 86.895   130.658   157.372   115.02      0.0      61.1441  107.684 
 79.6504   90.9302  102.107    92.8732   73.7605    0.0      52.4195
 27.83     47.5529   56.5124   41.0527  119.529    52.1964    0.0   

In [5]:
tpath

8-element Array{String,1}:
 "North Adams, MA 01247, USA"
 "Fitchburg, MA 01420, USA"  
 "Lowell, MA, USA"           
 "Boston, MA, USA"           
 "Fall River, MA, USA"       
 "Worcester, MA, USA"        
 "Springfield, MA, USA"      
 "North Adams, MA 01247, USA"

**( b ) Add the following towns: Amherst, Braintree, Lawrence, Newton, Natick, Pittsfield, Plymouth, Provincetown, Salem, Woburn to the list above. Produce an array of distances in alphabetical order.**

In [6]:
A = render_map("map_q1.json");
B = render_map("map_q2.json");
C = render_map("map_q3.json");
D = render_map("map_q4.json");

In [7]:
the_map = 0.000621371*[A C; B D]

17×17 Array{Float64,2}:
   0.0      47.0539   83.0232   31.4153  …  162.206    59.7454   44.3963
  52.9017    0.0      51.8894   30.3484     115.131    22.1519   11.857 
  84.1908   52.3747    0.0      77.0351     103.698    73.246    62.9517
  33.1967   30.121    77.0469    0.0        147.455    32.996    17.6469
  86.895   130.658   157.372   115.02       245.81    143.349   128.0   
  79.6504   90.9302  102.107    92.8732  …  195.148   108.905    93.5561
  27.83     47.5529   56.5124   41.0527     155.498    65.5279   50.1782
  49.5929   93.6257  104.802    77.7174     197.843   106.047    90.6978
  61.0733   13.3582   42.8976   44.9941     103.208    34.2301   23.9352
  44.6225   29.6295   80.8273   16.7907     144.068    32.8457   22.3047
  43.8632   10.9796   52.9557   27.7846  …  123.364    34.9931   19.644 
  44.7151   24.1247   54.992    34.5731     125.4      41.7816   26.4319
  99.6381  138.095   149.271   140.037      242.312   156.069   140.72  
  88.2838   40.5339   37.58

In [8]:
#run_TSP(the_map)

This doesn't work due to lack of resources, requires way (way) more than the average computer.  
17! permutations...uhh...no.

**( c ) Use a Simulated Annealing algorithm to find a minimum path through all of the towns in `(a)` and `(b)`. Use as large a number of randomizations in your algorithm as possible. Find the total distance as well as the path of the towns. Start and end with Fitchburg and write down the town names of the shortest path.**

In [9]:
rpath=nthperm(collect(1:17),rand(1:factorial(4)))

17-element Array{Int64,1}:
  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
 11
 12
 13
 16
 15
 14
 17

In [10]:
current_dist=total_distance(rpath, the_map)

1244.490537994

In [11]:
swapEls!(rpath,1,4)

17-element Array{Int64,1}:
  4
  2
  3
  1
  5
  6
  7
  8
  9
 10
 11
 12
 13
 16
 15
 14
 17

In [12]:
new_path = swapEls!(rpath,rand(1:17),rand(1:17))

17-element Array{Int64,1}:
  4
  2
  3
 11
  5
  6
  7
  8
  9
 10
  1
 12
 13
 16
 15
 14
 17

In [13]:
total_distance(new_path,the_map)

1223.1594929350001

In [14]:
sim_annealing(the_map,10^8)

([14,13,12,10,5,6,8,1,16,11,3,7,15,17,9,2,4],709.5323602220002)

In [36]:
@time out = sim_annealing(the_map,10^8)

  9.714622 seconds (9 allocations: 480 bytes)


([9,10,12,7,3,1,5,2,4,8,6,11,13],14.358641068)

In [ ]:
map(a->out[1])

Hmm......

## Problem 2:
Consider the following weighted graph:
<img src="graph01.png" width=250 />
Use the method from class using tropical arithmetic on matrices to find the shortest path from A to H.

In [16]:
w = [0   9   Inf 13  16  Inf Inf Inf;
     9   0   15  Inf 3   Inf Inf Inf;
     Inf 15  0   Inf 10  4   Inf Inf;
     13  Inf Inf 0   18  Inf Inf Inf;
     16  3   10  18  0   17  21  Inf;
     Inf Inf 4   Inf 17  0   12  24;
     Inf Inf Inf Inf 21  12  0   5;
     Inf Inf Inf Inf Inf 24  5   0]

8×8 Array{Float64,2}:
   0.0    9.0  Inf     13.0   16.0  Inf    Inf    Inf  
   9.0    0.0   15.0  Inf      3.0  Inf    Inf    Inf  
 Inf     15.0    0.0  Inf     10.0    4.0  Inf    Inf  
  13.0  Inf    Inf      0.0   18.0  Inf    Inf    Inf  
  16.0    3.0   10.0   18.0    0.0   17.0   21.0  Inf  
 Inf    Inf      4.0  Inf     17.0    0.0   12.0   24.0
 Inf    Inf    Inf    Inf     21.0   12.0    0.0    5.0
 Inf    Inf    Inf    Inf    Inf     24.0    5.0    0.0

In [17]:
using TropicalMath

In [18]:
w⊗w

8×8 Array{Float64,2}:
   0.0    9.0  24.0   13.0  12.0  33.0  37.0  Inf  
   9.0    0.0  13.0   21.0   3.0  19.0  24.0  Inf  
  24.0   13.0   0.0   28.0  10.0   4.0  16.0   28.0
  13.0   21.0  28.0    0.0  18.0  35.0  39.0  Inf  
  12.0    3.0  10.0   18.0   0.0  14.0  21.0   26.0
  33.0   19.0   4.0   35.0  14.0   0.0  12.0   17.0
  37.0   24.0  16.0   39.0  21.0  12.0   0.0    5.0
 Inf    Inf    28.0  Inf    26.0  17.0   5.0    0.0

In [19]:
1.0*((w⊗w)⊗w)⊗w

8×8 Array{Float64,2}:
  0.0   9.0  22.0  13.0  12.0  26.0  33.0  38.0
  9.0   0.0  13.0  21.0   3.0  17.0  24.0  29.0
 22.0  13.0   0.0  28.0  10.0   4.0  16.0  21.0
 13.0  21.0  28.0   0.0  18.0  32.0  39.0  44.0
 12.0   3.0  10.0  18.0   0.0  14.0  21.0  26.0
 26.0  17.0   4.0  32.0  14.0   0.0  12.0  17.0
 33.0  24.0  16.0  39.0  21.0  12.0   0.0   5.0
 38.0  29.0  21.0  44.0  26.0  17.0   5.0   0.0

^ These are the shortest distances with a max step size of 3  
The shortest distance from A => H is 42

## Problem 3:
Use the Dijkstra’s Algorithm to find the shortest path from A to H for the graph in the previous problem.

In [20]:
using Graphs;

In [21]:
g = simple_inclist(8,is_directed=false)

Undirected Graph (8 vertices, 0 edges)

In [22]:
A = 1
B = 2
C = 3
D = 4
E = 5
F = 6
G = 7
H = 8
inputs = [
    (A, B, 9.0),
    (A, E, 16.0),
    (A, D, 13.0),
    (B, A, 9.0),
    (B, E, 3.0),
    (B, C, 15.0),
    (C, B, 15.0),
    (C, E, 10.0),
    (C, F, 4.0),
    (D, A, 13.0),
    (D, E, 18.0),
    (E, D, 18.0),
    (E, A, 16.0),
    (E, B, 3.0),
    (E, C, 10.0),
    (E, F, 17.0),
    (E, G, 21.0),
    (F, C, 4.0),
    (F, E, 17.0),
    (F, G, 12.0),
    (F, H, 24.0),
    (G, E, 21.0),
    (G, F, 12.0),
    (G, H, 5.0),
    (H, F, 24.0),
    (H, G, 5.0)
]

26-element Array{Tuple{Int64,Int64,Float64},1}:
 (1,2,9.0) 
 (1,5,16.0)
 (1,4,13.0)
 (2,1,9.0) 
 (2,5,3.0) 
 (2,3,15.0)
 (3,2,15.0)
 (3,5,10.0)
 (3,6,4.0) 
 (4,1,13.0)
 (4,5,18.0)
 (5,4,18.0)
 (5,1,16.0)
 (5,2,3.0) 
 (5,3,10.0)
 (5,6,17.0)
 (5,7,21.0)
 (6,3,4.0) 
 (6,5,17.0)
 (6,7,12.0)
 (6,8,24.0)
 (7,5,21.0)
 (7,6,12.0)
 (7,8,5.0) 
 (8,6,24.0)
 (8,7,5.0) 

In [23]:
dist = map(a->a[3],inputs)

26-element Array{Float64,1}:
  9.0
 16.0
 13.0
  9.0
  3.0
 15.0
 15.0
 10.0
  4.0
 13.0
 18.0
 18.0
 16.0
  3.0
 10.0
 17.0
 21.0
  4.0
 17.0
 12.0
 24.0
 21.0
 12.0
  5.0
 24.0
  5.0

In [24]:
for edge in inputs
    add_edge!(g,edge[1],edge[2])
end

In [25]:
adjacency_matrix(g)

8×8 Array{Bool,2}:
 false   true  false   true   true  false  false  false
  true  false   true  false   true  false  false  false
 false   true  false  false   true   true  false  false
  true  false  false  false   true  false  false  false
  true   true   true   true  false   true   true  false
 false  false   true  false   true  false   true   true
 false  false  false  false   true   true  false   true
 false  false  false  false  false   true   true  false

In [26]:
r = dijkstra_shortest_paths(g, dist, 1)

Graphs.DijkstraStates{Int64,Float64,DataStructures.MutableBinaryHeap{Graphs.DijkstraHEntry{Int64,Float64},DataStructures.LessThan},Int64}([1,1,5,1,2,3,5,7],[1,1,5,1,2,3,5,7],[0.0,9.0,22.0,13.0,12.0,26.0,33.0,38.0],[2,2,2,2,2,2,2,2],MutableBinaryHeap(),[0,1,4,3,2,5,6,7])

In [27]:
r.dists

8-element Array{Float64,1}:
  0.0
  9.0
 22.0
 13.0
 12.0
 26.0
 33.0
 38.0

^ These are the shortest distances from A to each letter.
- A => A is 0
- A => B is 9
- A => C is 22
- A => D is 13  
.  
.  
.  
- A => H is 38

Which are all the same shortest paths as the tropical math solution

### Problem 4:
This problem finds the shortest distance from Fitchburg State to the Entertainment Cinemas in Leominster
which mimics Google Maps or another mapping/GPS program.

**( a ) Create a graph that has Fitchburg State as the starting point (use 160 Pearl St.) and the Entertainment
Cinemas (45 Sack Blvd, Leominster, MA 01453) as the end. Include the following intersections in your
graph:**
- North Street and Main Street (Fitchburg)
- John Fitch Hwy and Lunenburg Street (Fitchburg)
- South Street and Laurel Street (Fitchburg)
- South Street and Wanoosnoc Rd (Fitchburg)
- Water Street and Bemis Rd (Fitchburg)
- Summer Street and John Fitch Hwy (Fitchburg)
- North Street and Main Street (Lunenburg)
- Main Street and Prospect Street (Lunenburg)
- N. Main Street and Rte 2 (Lunenburg)
- Merriam Ave and Rte 2 (Lunenburg)
- Haws Street and Commercial Road (Lunenburg)

In [28]:
A = render_map("cinemas_q1.json");
B = render_map("cinemas_q2.json");
C = render_map("cinemas_q3.json");
D = render_map("cinemas_q4.json");
the_map = 0.000621371*[A C; B D]

13×13 Array{Float64,2}:
 0.0       0.472863  0.695314  0.818346  …  5.15614   5.46869   5.85145 
 0.472863  0.0       0.840094  0.300122     4.63853   4.95046   5.33323 
 0.695314  0.840094  0.0       1.14084      4.68141   4.99396   5.37672 
 0.797219  0.300744  1.14084   0.0          4.47449   5.23008   5.61284 
 2.55881   2.06171   2.28665   1.76097      2.71291   3.46911   3.85188 
 2.35313   1.64601   1.6572    1.61743   …  3.04161   3.35354   3.73693 
 1.77526   1.11909   1.07932   1.42667      3.60706   3.91961   4.30237 
 3.39393   2.73714   2.69799   3.04472      3.25785   2.32144   2.70421 
 1.17253   0.758694  2.83967   0.873648     5.34752   5.47987   5.86264 
 4.78331   4.28684   4.34028   4.1986       0.361638  0.853764  0.682887
 5.13936   4.64226   4.6957    4.56956   …  0.0       1.10045   1.01159 
 5.8558    5.3587    4.82867   5.05796      1.19552   0.0       0.382765
 6.23856   5.74147   5.21144   5.44135      1.03272   0.382765  0.0     

**( b ) Use the Tropical Matrix Arithmetic method to solve this problem. List the path travelled that minimizes the travel distance and the nodes (intersections) travelled through.**

In [29]:
the_map = (the_map⊗the_map)⊗the_map

13×13 Array{Float64,2}:
 0.0       0.472863  0.695314  0.772986  …  5.11078   5.42333   5.43203 
 0.472863  0.0       0.840094  0.300122     4.63791   4.95046   4.95916 
 0.695314  0.840094  0.0       1.14022      4.68141   4.99396   5.00266 
 0.773607  0.300744  1.14084   0.0          4.47387   4.88149   4.89081 
 2.53457   2.06171   2.28602   1.76097      2.71291   3.46911   3.65801 
 2.11888   1.64601   1.6572    1.61743   …  3.04161   3.35354   3.36286 
 1.59195   1.11909   1.07932   1.41921      3.60706   3.91961   3.92831 
 3.21      2.73714   2.69799   3.03726      3.25785   2.32144   2.70421 
 1.17253   0.758694  1.59879   0.873648     5.16732   5.47987   5.48857 
 4.7597    4.28684   4.34028   4.1986       0.361638  0.853764  0.682887
 5.11513   4.64226   4.6957    4.55403   …  0.0       1.10045   1.01159 
 5.22262   4.74976   4.76094   4.72118      1.19552   0.0       0.382765
 5.44259   4.96973   5.02316   4.88149      1.03272   0.382765  0.0     

The shortest distance from *Fitchburg State* to *Entertainment Cinemas* is 5.43 miles

**( c ) Use Dijkstra’s method to solve the problem. List the path travelled that minimizes the travel distance and the nodes (intersections) travelled through.**

In [30]:
A = 1   # 160 Pearl Street, Fitchburg MA (Fitchburg State)
B = 2   # Main St & North Street, Fitchburg MA
C = 3   # Lunenburg St & John Fitch Hwy, Fitchburg MA
D = 4   # Laurel St & South St, Fitchburg MA
E = 5   # South St & Wanoosnoc Rd, Fitchburg MA
F = 6   # Water St & Bemis Rd, Fitchburg MA
G = 7   # John Fitch Hwy & Summer St, Fitchburg MA
H = 8   # Main St & North St, Leominster MA
I = 9   # Main Street & Prospect St, Fitchburg MA
J = 10  # 2 N Main St, Leominster MA
K = 11  # 2 Merriam Ave, Leominster MA
L = 12  # Commercial Rd & Haws St, Leominster MA
M = 13  # 45 Sack Blvd, Leominster MA (Entertainment Cinemas)

# didn't end up using the above, automated filling the cinema_map array like a smart person.

cinema_map = Array{Tuple{Int, Int, Float64}}(0)

for i=1:13
    for j=1:13
        if the_map[i,j] != 0
            push!(cinema_map,(i,j,the_map[i,j]))
        end
    end
end
#the_map[1,:]
cinema_map

156-element Array{Tuple{Int64,Int64,Float64},1}:
 (1,2,0.472863)  
 (1,3,0.695314)  
 (1,4,0.772986)  
 (1,5,2.53395)   
 (1,6,2.06979)   
 (1,7,1.59817)   
 (1,8,3.20814)   
 (1,9,1.25952)   
 (1,10,4.74914)  
 (1,11,5.11078)  
 (1,12,5.42333)  
 (1,13,5.43203)  
 (2,1,0.472863)  
 ⋮               
 (13,1,5.44259)  
 (13,2,4.96973)  
 (13,3,5.02316)  
 (13,4,4.88149)  
 (13,5,3.67976)  
 (13,6,3.36597)  
 (13,7,4.04947)  
 (13,8,2.71353)  
 (13,9,5.565)    
 (13,10,0.682887)
 (13,11,1.03272) 
 (13,12,0.382765)

In [31]:
dist_cinema_map = map(a->a[3],cinema_map)

156-element Array{Float64,1}:
 0.472863
 0.695314
 0.772986
 2.53395 
 2.06979 
 1.59817 
 3.20814 
 1.25952 
 4.74914 
 5.11078 
 5.42333 
 5.43203 
 0.472863
 ⋮       
 5.44259 
 4.96973 
 5.02316 
 4.88149 
 3.67976 
 3.36597 
 4.04947 
 2.71353 
 5.565   
 0.682887
 1.03272 
 0.382765

In [32]:
g2 = simple_inclist(13,is_directed=false)
for edge in cinema_map
    add_edge!(g2,edge[1],edge[2])
end
g2

Undirected Graph (13 vertices, 156 edges)

In [33]:
adjacency_matrix(g2)

13×13 Array{Bool,2}:
 false   true   true   true   true  …   true   true   true   true   true
  true  false   true   true   true      true   true   true   true   true
  true   true  false   true   true      true   true   true   true   true
  true   true   true  false   true      true   true   true   true   true
  true   true   true   true  false      true   true   true   true   true
  true   true   true   true   true  …   true   true   true   true   true
  true   true   true   true   true      true   true   true   true   true
  true   true   true   true   true      true   true   true   true   true
  true   true   true   true   true     false   true   true   true   true
  true   true   true   true   true      true  false   true   true   true
  true   true   true   true   true  …   true   true  false   true   true
  true   true   true   true   true      true   true   true  false   true
  true   true   true   true   true      true   true   true   true  false

In [34]:
r2 = dijkstra_shortest_paths(g2, dist_cinema_map, 1)

Graphs.DijkstraStates{Int64,Float64,DataStructures.MutableBinaryHeap{Graphs.DijkstraHEntry{Int64,Float64},DataStructures.LessThan},Int64}([1,1,1,1,1,1,1,1,1,1,10,6,1],[1,1,1,1,1,1,1,1,1,1,10,6,1],[0.0,0.472863,0.695314,0.772986,2.53395,2.06979,1.59195,3.20814,1.17253,4.74914,5.10456,5.17353,5.43203],[2,2,2,2,2,2,2,2,2,2,2,2,2],MutableBinaryHeap(),[0,1,2,3,4,5,6,7,8,9,10,11,12])

In [35]:
r2.dists

13-element Array{Float64,1}:
 0.0     
 0.472863
 0.695314
 0.772986
 2.53395 
 2.06979 
 1.59195 
 3.20814 
 1.17253 
 4.74914 
 5.10456 
 5.17353 
 5.43203 